receiver/prometheusremotewritereceiver/receiver.go

```diff
@@ -0,0 +1,243 @@
+			defer rec.shutdownWG.Done()
+			if errHTTP := rec.server.Serve(listener); errHTTP != http.ErrServerClosed {
+				host.ReportFatalError(errHTTP)
+			}
+		}()
+	})
+	return err
+}
+
+func (rec *PrometheusRemoteWriteReceiver) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	req, err := remote.DecodeWriteRequest(r.Body)
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	ctx := rec.obsrecv.StartMetricsOp(context.Background())
+
+	reg := regexp.MustCompile(`(\w+)_(\w+)_(\w+)\z`)
+	timeThreshold := time.Now().Add(-time.Hour * 24)
+	pms := pmetric.NewMetrics()
+
+	for _, ts := range req.Timeseries {
+
+		prm := pmetric.NewResourceMetrics()
+
+		metricName := finaName(ts.Labels)
+		pm := pmetric.NewMetric()
+		pm.SetName(metricName)
+		rec.logger.Debug("Metric name", zap.String("metric_name", pm.Name()))
+
+		match := reg.FindStringSubmatch(metricName)
+		metricsType := ""
+		if len(match) > 1 {
+			lastSuffixInMetricName := match[len(match)-1]
+			if IsValidSuffix(lastSuffixInMetricName) {
+				metricsType = lastSuffixInMetricName
+				if len(match) > 2 {
+					secondSuffixInMetricName := match[len(match)-2]
+					if IsValidUnit(secondSuffixInMetricName) {
+						pm.SetUnit(secondSuffixInMetricName)
+					}
+				}
+			} else if IsValidUnit(lastSuffixInMetricName) {
+				pm.SetUnit(lastSuffixInMetricName)
+			}
+		}
+		rec.logger.Debug("Metric unit", zap.String("metric name", pm.Name()), zap.String("metric_unit", pm.Unit()))
+
+		for _, s := range ts.Samples {
+			ppoint := pmetric.NewNumberDataPoint()
+			ppoint.SetDoubleValue(s.Value)
+			ppoint.SetTimestamp(pcommon.NewTimestampFromTime(time.Unix(0, s.Timestamp*int64(time.Millisecond))))
+			if ppoint.Timestamp().AsTime().Before(timeThreshold) {
+				rec.logger.Debug("Metric older than the threshold", zap.String("metric name", pm.Name()), zap.Time("metric_timestamp", ppoint.Timestamp().AsTime()))
+				continue
+			}
+			for _, l := range ts.Labels {
+				labelName := l.Name
+				if l.Name == "__name__" {
+					labelName = "key_name"
+				}
+				ppoint.Attributes().PutString(labelName, l.Value)
+			}
+			if IsValidCumulativeSuffix(metricsType) {
+				pm.SetEmptySum()
+				pm.Sum().SetIsMonotonic(true)
+				pm.Sum().SetAggregationTemporality(pmetric.MetricAggregationTemporalityCumulative)
+				ppoint.CopyTo(pm.Sum().DataPoints().AppendEmpty())
+			} else {
+				pm.SetEmptyGauge()
+				ppoint.CopyTo(pm.Gauge().DataPoints().AppendEmpty())
+			}
+			rec.logger.Debug("Metric sample",
+				zap.String("metric_name", pm.Name()),
+				zap.String("metric_unit", pm.Unit()),
+				zap.Float64("metric_value", ppoint.DoubleValue()),
+				zap.Time("metric_timestamp", ppoint.Timestamp().AsTime()),
+				zap.String("metric_labels", fmt.Sprintf("%#v", ppoint.Attributes())),
+			)
+		}
+		pilm := pmetric.NewScopeMetrics()
+		pm.CopyTo(pilm.Metrics().AppendEmpty())
+		pilm.CopyTo(prm.ScopeMetrics().AppendEmpty())
+		prm.CopyTo(pms.ResourceMetrics().AppendEmpty())
```

Comment:
```suggestion
	pm.MoveAndAppendTo(pms.ResourceMetrics().AppendEmpty().ScopeMetrics().AppendEmpty().Metrics())
```

Not sure that creating a new `ResourceMetrics` for every timeseries is appropriate, and there will need to be work done to ensure that the resource and scope attributes are properly extracted, but if this is the way it'll be done best to avoid all the copying.  Better yet, do this all up at the top of this loop:

```go
pm := pms.ResourceMetrics().AppendEmpty().ScopeMetrics().AppendEmpty().Metrics().AppendEmpty()
```

and then fill it in from there.

Reply:
done